### PR TITLE
[Fit API] Update fit method and LoggingHandler

### PR DESCRIPTION
**Pull request by abhinavs95**
## Description ##
- moved train loop outside `fit()` into a new `train_epoch()` method. This allows tasks with custom train loop to use Estimator by simply overriding the `train_epoch()` method.
- update precision for learning rate in LoggingHandler to accurately display low values
- fix reading optimizer name from trainer object in LoggingHandler

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

@roywei @nswamy

### CODE REVIEW

**Comment by karan6181** [comment]
@mxnet-label-bot add [Gluon, pr-awaiting-review]

**Comment by piyushghai** [comment]
@abhinavs95 Can you look into the review comments ?

**Comment by vandanavk** [comment]
@mxnet-label-bot update [Gluon, pr-awaiting-response]

**Comment by abhinavs95** [comment]
@szha For my use case, I want perform certain operations (like gradient clipping) in between batch-begin and batch-end which cannot be captured by event handlers. Also my model is multi-input, multi-output so I also have modified the forward pass and loss computation. To use estimator for cases like these having the train method out of fit() is necessary so that the user can inject their custom code. 

Also this does not break the basic functionality of estimator as the default train method is used if the user does not specify their own.

**Comment by szha** [comment]
@abhinavs95 `train_epoch`'s granularity is too coarse to ensure that registered handlers work correctly.

**Comment by abhinavs95** [comment]
Closing this for now, will reopen when I get a chance to work on it.
